Set object lock for volumes for cephfs encryption

Sunnatillo opened the pull request:
<!-- Please take a look at our [Contributing](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#Code-contribution-workflow)
documentation before submitting a Pull Request!
Thank you for contributing to ceph-csi! -->

# Describe what this PR does #

Provide some context for the reviewer

## Is there anything that requires special attention ##

Do you have any questions?

Is the change backward compatible?

Are there concerns around backward compatibility?

Provide any external context for the change, if any.

For example:

* Kubernetes links that explain why the change is required
* CSI spec related changes/catch-up that necessitates this patch
* golang related practices that necessitates this change

## Related issues ##

Mention any github issues relevant to this PR. Adding below line
will help to auto close the issue once the PR is merged.

Closes: #4688
Fixes: #4654

## Future concerns ##

List items that are not part of the PR and do not impact it's
functionality, but are work items that can be taken up subsequently.

**Checklist:**

* [ ] **Commit Message Formatting**: Commit titles and messages follow
  guidelines in the [developer
  guide](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#commit-messages).
* [ ] Reviewed the developer guide on [Submitting a Pull
  Request](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#development-workflow)
* [ ] [Pending release
  notes](https://github.com/ceph/ceph-csi/blob/devel/PendingReleaseNotes.md)
  updated with breaking and/or notable changes for the next major release.
* [ ] Documentation has been updated, if necessary.
* [ ] Unit tests have been added, if necessary.
* [ ] Integration tests have been added, if necessary.

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

* `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)

</details>

nixpanic commented on internal/cephfs/nodeserver.go:
```diff
@@ -127,13 +135,72 @@ func maybeUnlockFileEncryption(
+		return err
+	}
+
+	if created {
```
Usually we write code in a way that handles errors or unexpected results, and bails out, only when everything succeeds it continues in the function.

```go
if !created {
    return fmt.Errorf("There is already one file system with name %s", string(volID))
}

defer func() {
...
```

Sunnatillo replied:
Done

nixpanic reviewed (commented):
just a nit to improve the readability of the code a little, looks quite good already

Sunnatillo committed:
cephfs: Set object lock for volumes for cephfs encryption

The way fscrypt client handles metadata and policy creation
causing errors when multiple instances start simultaneously.
This commit adds a lock to ensure the initial setup
completes correctly, preventing race conditions and
mismatches.

Signed-off-by: Sunnatillo <sunnat.samadov@est.tech>

Madhu-1 commented on internal/cephfs/nodeserver.go:
```diff
@@ -127,13 +128,47 @@ func maybeUnlockFileEncryption(
+	log.DebugLog(ctx, "Creating lock for the following volume ID %s", volID)
+
+	ioctx, err := volOptions.GetConnection().GetIoctx(volOptions.MetadataPool)
+	if err != nil {
```
return error from here?

Sunnatillo replied:
Hey. Thank you for the review. I have addressed most of your comments.

nixpanic replied:
if creation of the ioctx fails, an error should be returned

nixpanic replied:
@Sunnatillo, this is the last error that still needs to be handled. If getting the ioctx fails, there is no need to continue.

Madhu-1 commented on internal/cephfs/nodeserver.go:
```diff
@@ -127,13 +128,47 @@ func maybeUnlockFileEncryption(
+	}
+	defer ioctx.Destroy()
+
+	res, err := ioctx.LockExclusive(volOptions.VolID, lockName, string(lockCookie), lockDesc, lockDuration, &flags)
```
```suggestion
res, err := ioctx.LockExclusive(volOptions.VolID, lockName, lockCookie, lockDesc, lockDuration, &flags)
```
lockCookie is already a string

Madhu-1 commented on internal/cephfs/nodeserver.go:
```diff
@@ -127,13 +128,47 @@ func maybeUnlockFileEncryption(
+	lockName := string(volID) + "-mutexLock"
+	lockCookie := lockName + "-coockie"
+	lockDesc := "Lock for " + string(volID)
+	lockDuration := 40 * time.Second
```
should we have some higher value for this one like (150S) default is 180S in kubelet? and make it configurable so user can adjust this one based on the network connectivity with ceph cluster?

Sunnatillo replied:
Should we add flag to make it configurable?

nixpanic replied:
My preference is to have a high timeout (150 seconds like it is now should be plenty). Adding a configuration option if really needed can be done later. But hopefully this is sufficient, I expect users that need to have a higher timeout to have more severe problems than failures to setup encryption.

Madhu-1 replied:
works for me as well

Madhu-1 commented on internal/cephfs/nodeserver.go:
```diff
@@ -127,13 +128,47 @@ func maybeUnlockFileEncryption(
+		log.DebugLog(ctx, "Releasing following lock  %s", lockName)
 
-	return nil
+		_, err = ioctx.Unlock(string(volID), lockName, lockCookie)
```
wont we get into the problem if we try to unlock if someone else is already holding the lock? in case of RWX and multiple nodes

Madhu-1 commented on internal/cephfs/nodeserver.go:
```diff
@@ -127,13 +128,47 @@ func maybeUnlockFileEncryption(
+
+	if res == 0 {
+		log.DebugLog(ctx, "cephfs: unlocking fscrypt on volume %q path %s", volID, stagingTargetPath)
 		return fscrypt.Unlock(ctx, volOptions.Encryption, stagingTargetPath, string(volID))
```
dont we need to release lock once the fscrypt.Unlock is done?

Madhu-1 commented on internal/cephfs/nodeserver.go:
```diff
@@ -127,13 +128,47 @@ func maybeUnlockFileEncryption(
+		}
+	}()
+
+	return fmt.Errorf("There is already one file system with name %s", string(volID))
```
Need to update the error message

Madhu-1 commented:
@Sunnatillo can you please check the CI failures?

Madhu-1 commented on internal/cephfs/nodeserver.go:
```diff
@@ -127,10 +133,55 @@ func maybeUnlockFileEncryption(
+	defer ioctx.Destroy()
+
+	res, err := ioctx.LockExclusive(volOptions.VolID, lockName, lockCookie, lockDesc, lockDuration, &flags)
+	switch res {
```
can we make use of https://pkg.go.dev/syscall? and also need to return errors instead of returning nil until the fscrypt.Unlock is done or how to ensure that unlock is successful?

Madhu-1 replied:
i think we might need to make use of metadata to ensure it's in progress or is done along with the lock.

nixpanic replied:
I do not see how the syscall package can be used here.

Sunnatillo replied:
Thank you for the reviews @Madhu-1. 

Sorry, I might misunderstood you.  `fscrypt.Unlock` returns an error if it is not successful. 

Could you please elaborate more regarding https://pkg.go.dev/syscall?

Madhu-1 replied:
i mean like below

```go
if errors.Is(err, syscall.EEXIST) {
```

if we do like above we don't need to import c to this file

nixpanic commented on internal/cephfs/nodeserver.go:
```diff
@@ -127,10 +133,55 @@ func maybeUnlockFileEncryption(
+		return fmt.Errorf("Failed to lock volume ID %v: %w", volID, err)
+	}
+
+	if res == 0 {
```
instead of positive check, use
```go
if res != 0 {
    return err // can this happen? err == nil
}

// Success! Continue with the function :)
```

nixpanic replied:
@Sunnatillo please check the above to improve readability of the code.

Sunnatillo replied:
Thank you for the review @nixpanic. 

`res` can be equal to `0`, `C.EBUSY` and `-C.EEXIST`. Only in the case of  `res == 0` it should try to unlock fscrypt on volume.  In my opinion `if res != 0 ` is not suitable here.

nixpanic replied:
@Sunnatillo , the thing I wanted to suggest is to return early from the function when `res != 0`. In the case where `res == 0` the function can continue (and you can drop one tab indention). I think this makes the code a little easier to understand.

nixpanic commented on internal/cephfs/nodeserver.go:
```diff
@@ -127,10 +133,55 @@ func maybeUnlockFileEncryption(
+		err = fscrypt.Unlock(ctx, volOptions.Encryption, stagingTargetPath, string(volID))
+		if err != nil {
+			return err
+		} else {
```
no need for the `else {`, just continue with the function and loose one indention level (makes it easier to read and understand the code).

nixpanic commented on internal/cephfs/nodeserver.go:
```diff
@@ -127,10 +129,54 @@ func maybeUnlockFileEncryption(
+		default:
+			return fmt.Errorf("Failed to lock volume ID %v: %w", volID, err)
+		}
+		return err
```
**golangci-lint** wants an empty line before this return statement :disappointed:

Madhu-1 commented on internal/cephfs/nodeserver.go:
```diff
@@ -127,10 +129,54 @@ func maybeUnlockFileEncryption(
+	defer ioctx.Destroy()
+
+	res, err := ioctx.LockExclusive(volOptions.VolID, lockName, lockCookie, lockDesc, lockDuration, &flags)
+	if res != 0 {
```
https://github.com/ceph/go-ceph/blob/1046b034a1f618f67acd3c6523482917e27c7113/rados/ioctx.go#L545-L558 Looking at this one, we need to return an error in case of EBUSY and EEXIST at line 161 it return nil, we need to have return in those cases as well, @nixpanic can you please check this one?

nixpanic replied:
EBUSY and EEXIST are expected return values, not an error while calling the function. I do not think go-ceph should return an error in those cases.

Madhu-1 replied:
sorry i meant not in the go-ceph from here cephcsi should return an error so that if the lock is held on one node we should keep retrying on another node until it's done, or we just return nil (which is happening here) assuming that the other client will unlock fscrypt?

nixpanic replied:
ah, yes, now I see what you mean. If EBUSY/EEXISTS is returned from `LockExclusive()`, `err` will be `nil`. That `err` is then returned on line 162, and the caller of this function assumes all is OK (because `err` is `nil`). This is not correct, and instead of returning `err` on line 162, a new error should be returned. Ideally each `case` in the below switch sets `err` to a new unique value (mentioning something similar as what is in the debug log).

nixpanic commented on Makefile:
```diff
@@ -162,7 +162,7 @@ commitlint:
 cephcsi: check-env
 	if [ ! -d ./vendor ]; then (go mod tidy && go mod vendor); fi
-	GOOS=linux go build $(GO_TAGS) -mod vendor -a -ldflags '$(LDFLAGS)' -o _output/cephcsi ./cmd/
+	GOOS=linux go build $(GO_TAGS) -buildvcs=false -mod vendor -a -ldflags '$(LDFLAGS)' -o _output/cephcsi ./cmd/
```
this now got added, is that intentional?

Sunnatillo replied:
My bad.  it is not intentional

Sunnatillo replied:
removed now.

Madhu-1 commented on internal/cephfs/nodeserver.go:
```diff
@@ -127,15 +129,70 @@ func maybeUnlockFileEncryption(
+
+	ioctx, err := volOptions.GetConnection().GetIoctx(volOptions.MetadataPool)
+	if err != nil {
+		log.ErrorLog(ctx, "Failed to create ioctx: %s", err)
```
return err is missing here, everything else LGTM

Sunnatillo replied:
added

nixpanic replied:
Thanks for your persistence!

Madhu-1 reviewed (dismissed):
Thank you @Sunnatillo 🎉

nixpanic commented:
@mergifyio rebase

Sunnatillo commented:
Thank you for reviews @nixpanic, @Madhu-1. 

I rebased to latest devel branch.

iPraveenParihar commented:
@nixpanic, I see `fscrypt.Unlock` call for RBD fscrypt. Do we need to do the same here as well?
do we need to
https://github.com/ceph/ceph-csi/blob/e71a95fece24de8de8f1de51201a7314af4f3480/internal/rbd/nodeserver.go#L467-L474

nixpanic commented:
> @nixpanic, I see `fscrypt.Unlock` call for RBD fscrypt. Do we need to do the same here as well?

No, for RBD that is not needed. RBD with a filesystem is never attached to more than one workernode at the same time, so initialization/unlocking of the filesystem will always be on a single node. It is different with ReadWriteMany RBD volumes in block-mode with encryption. The LUKS calls would benefit from a similar locking.

iPraveenParihar reviewed (approved):
Thanks!

NymanRobin commented on internal/cephfs/nodeserver.go:
```diff
@@ -127,15 +129,72 @@ func maybeUnlockFileEncryption(
+
+	log.DebugLog(ctx, "cephfs: unlocking fscrypt on volume %q path %s", volID, stagingTargetPath)
+	err = fscrypt.Unlock(ctx, volOptions.Encryption, stagingTargetPath, string(volID))
+	if err != nil {
```
Won’t this leave a hanging lock in case the fscrypt fails to unlock the filesystem?

I think the ”ceph unlock functionality” could be moved to a separate function and this function pushed to the defer stack after the exclusiveLock is successful?

Sunnatillo replied:
What about `lockDuration` variable on the `LockExclusive` function? When duration expires it will automatically unlock?

NymanRobin replied:
Yes, this I understand. But it cause unnecessary delays in my opinion that could easily be avoided. My point is that usually this is operation that would take seconds start to take minutes due to one failure

Sunnatillo replied:
@nixpanic, @Madhu-1 WDYT?

Madhu-1 replied:
Yes make sense to avoid unwanted delay in this case,As suggested `ioctx.Unlock` can be called defer right after the successful Lock.

NymanRobin replied:
I created an issue for this so it is not forgotten: https://github.com/ceph/ceph-csi/issues/4715

NymanRobin reviewed (commented):
Thanks a lot for fixing this 🎉

But I have a slight concern in the scenario when facrypt.Unlock fails. What do you think?

z2000l commented:
Is there a config to use the lock? We built an image from devel branch (https://github.com/ceph/ceph-csi/tree/devel) and tried it out. However pods failed to come up and we got the following error:

Warning  FailedMount             15s (x6 over 31s)  kubelet                  MountVolume.MountDevice failed for volume "pvc-4642810a-7f00-4ae0-8583-ddbcee6ed314" : rpc error: code = Internal desc = Failed to lock volume ID 0001-0009-rook-ceph-00000000000000ad-d71d2949-db34-434e-8008-5497dad5c70b: rados: ret=-1, Operation not permitted

NymanRobin commented:
I have tested this and encountered the same error as @z2000l. @Sunnatillo, could you please document any additional settings required or anything we might be overlooking?

NymanRobin commented on internal/cephfs/nodeserver.go:
```diff
@@ -127,15 +129,72 @@ func maybeUnlockFileEncryption(
+	}
+	defer ioctx.Destroy()
+
+	res, err := ioctx.LockExclusive(volOptions.VolID, lockName, lockCookie, lockDesc, lockDuration, &flags)
```
~~When reading the code in go-ceph it seems to be that the object that we want to lock should exist and this won't create a new object? Are we sure there exist an object in rados corresponding to `VolID`, my initial assumption is no but I am not 100% confident, without further investigation~~
```
// LockExclusive takes an exclusive lock on an object.
func (ioctx *IOContext) LockExclusive(oid, name, cookie, desc string, duration time.Duration, flags *byte) (int, error) {
```
Edit: The object seems not to be needed

NymanRobin replied:
Continuing my dig here, I added the following logging before the LockExclusive

```
_, err = ioctx.Stat(volOptions.VolID)
if err != nil {
  log.ErrorLog(ctx, "The error when object is not here: %s", err)
  log.ErrorLog(ctx, "Object %s does not exist", volOptions.VolID)
}
```

I get these logs:
```
E0724 09:03:43.715019       1 nodeserver.go:156] ID: 142 Req-ID: 0001-0009-rook-ceph-0000000000000001-489ccf33-13a5-40fc-8460-7dd866bc44de The error when object is not here: rados: ret=-2, No such file or directory
E0724 09:03:43.715051       1 nodeserver.go:157] ID: 142 Req-ID: 0001-0009-rook-ceph-0000000000000001-489ccf33-13a5-40fc-8460-7dd866bc44de Object csi-vol-489ccf33-13a5-40fc-8460-7dd866bc44de does not exist
E0724 09:03:43.715477       1 utils.go:240] ID: 142 Req-ID: 0001-0009-rook-ceph-0000000000000001-489ccf33-13a5-40fc-8460-7dd866bc44de GRPC error: rpc error: code = Internal desc = Failed to lock volume ID 0001-0009-rook-ceph-0000000000000001-489ccf33-13a5-40fc-8460-7dd866bc44de: rados: ret=-1, Operation not permitted
```
This makes me highly suspicious... I however wonder why in Ceph they do not return -2, in case a lock is tried to be acquired on a object that does not exist? Maybe they have not thought of this scenario and it could be reported?I guess next step in this journey to the core would be to create the object first and see if the lock can the be acquired :smile: 

Edit: Actual problem is comments below, the object seems to not be needed

NymanRobin replied:
Now I know at least why the -1 comes, and it is correct that the permissions are not there!
After adding the following to the authentication I can acquire the lock:
```
ceph auth caps client.csi-cephfs-node \
  mon 'allow *' \
  mgr 'allow *' \
  osd 'allow *' \
  mds 'allow *'
```

But I think what is actually needed is the following:
`ceph auth caps client.csi-cephfs-node rados 'allow rw'`

NymanRobin replied:
Okay to conclude this it seems the lock can be created without the object existing so that is all good and my ramblings from above can be ignored!

Except some more permission for the client.csi-cephfs-node  is needed to be able to create the lock

However it is extremely slow currently for me, bringing up five pods with 3 replicas and after 5 minutes I have only one replica of each. The error reported is the following:

```
  Warning  FailedMount             3m32s (x2 over 5m47s)  kubelet                  Unable to attach or mount volumes: unmounted volumes=[my-volume], unattached volumes=[my-volume kube-api-access-8cmvd]: timed out waiting for the condition
  Warning  FailedMount             81s (x11 over 7m35s)   kubelet                  MountVolume.MountDevice failed for volume "pvc-6afa784b-dc4f-44fb-bed9-8c6fe985c7e8" : rpc error: code = Internal desc = Lock is already held by another client and cookie pair for 0001-0009-rook-ceph-0000000000000001-c7f5d8f9-f9ca-4af8-8457-7ffd1bfb4437 volume
  Warning  FailedMount             74s                    kubelet                  Unable to attach or mount volumes: unmounted volumes=[my-volume], unattached volumes=[kube-api-access-8cmvd my-volume]: timed out waiting for the condition

```

This is the status of the pods after 8m46s:
```
$ kubectl get pods -n default 
NAME                            READY   STATUS              RESTARTS   AGE
deploy-pod-1-7f7789fd5f-92x4f   1/1     Running             0          8m46s
deploy-pod-1-7f7789fd5f-99qmp   1/1     Running             0          8m46s
deploy-pod-1-7f7789fd5f-dnr79   0/1     ContainerCreating   0          8m46s
deploy-pod-2-58c654874c-j2mvp   0/1     ContainerCreating   0          8m46s
deploy-pod-2-58c654874c-jh2wp   1/1     Running             0          8m46s
deploy-pod-2-58c654874c-svjws   1/1     Running             0          8m46s
deploy-pod-3-6674fc79f-rbtk2    1/1     Running             0          8m46s
deploy-pod-3-6674fc79f-w452p    1/1     Running             0          8m46s
deploy-pod-3-6674fc79f-w8wqb    0/1     ContainerCreating   0          8m46s
deploy-pod-4-6c958fb787-cqxlb   1/1     Running             0          8m46s
deploy-pod-4-6c958fb787-kvgbl   0/1     ContainerCreating   0          8m46s
deploy-pod-4-6c958fb787-q794c   1/1     Running             0          8m46s
deploy-pod-5-86b9bb47-5vmrq     0/1     ContainerCreating   0          8m45s
deploy-pod-5-86b9bb47-dd6v7     1/1     Running             0          8m45s
deploy-pod-5-86b9bb47-g5jvg     0/1     ContainerCreating   0          8m45s
```

After a whopping 13 minutes all the pods were finally running!

Should we actually poll the lock when trying to acquire, I think that might make it faster

Sunnatillo replied:
Thank you @NymanRobin for testing. 
I also tested. it is slow. 
It took me 8 minutes to 6 replicas to be ready. 
We should definitely improve it.

NymanRobin replied:
I can write up a ticket for the performance and for the permissions!

NymanRobin commented:
I can write up a ticket for the performance and for the permissions!
